src/cascadia/PublicTerminalCore/Source.cpp

```diff
@@ -0,0 +1,291 @@
+
+    auto dxEngine = std::make_unique<::Microsoft::Console::Render::DxEngine>();
+    RETURN_IF_FAILED(dxEngine->SetHwnd(_hwnd));
+    RETURN_IF_FAILED(dxEngine->Enable());
+    _renderer->AddRenderEngine(dxEngine.get());
+
+    auto pfn = std::bind(&::Microsoft::Console::Render::Renderer::IsGlyphWideByFont, _renderer.get(), std::placeholders::_1);
+    SetGlyphWidthFallback(pfn);
+
+    _UpdateFont(96);
+    RECT windowRect;
+    GetWindowRect(_hwnd, &windowRect);
+
+    const COORD windowSize{ static_cast<short>(windowRect.right - windowRect.left), static_cast<short>(windowRect.bottom - windowRect.top) };
+
+    // Fist set up the dx engine with the window size in pixels.
+    // Then, using the font, get the number of characters that can fit.
+    // Resize our terminal connection to match that size, and initialize the terminal with that size.
+    const auto viewInPixels = Viewport::FromDimensions({ 0, 0 }, windowSize);
+    RETURN_IF_FAILED(dxEngine->SetWindowSize({ viewInPixels.Width(), viewInPixels.Height() }));
+
+    _renderEngine = std::move(dxEngine);
+
+    _terminal->SetBackgroundCallback([](auto) {});
+
+    _terminal->Create(COORD{ 80, 25 }, 1000, *_renderer);
+    _terminal->SetDefaultBackground(RGB(5, 27, 80));
+    _terminal->SetDefaultForeground(RGB(255, 255, 255));
+
+    localPointerToThread->EnablePainting();
+
+    return S_OK;
+}
+
+void HwndTerminal::RegisterScrollCallback(std::function<void(int, int, int)> callback)
+{
+    _terminal->SetScrollPositionChangedCallback(callback);
+}
+
+void HwndTerminal::_UpdateFont(int newDpi)
+{
+    auto lock = _terminal->LockForWriting();
+
+    // TODO: MSFT:20895307 If the font doesn't exist, this doesn't
+    //      actually fail. We need a way to gracefully fallback.
+    _renderer->TriggerFontChange(newDpi, _desiredFont, _actualFont);
+}
+
+HRESULT HwndTerminal::Refresh(double width, double height, _Out_ int* charColumns, _Out_ int* charRows)
+{
+    auto lock = _terminal->LockForWriting();
+    const SIZE windowSize{ static_cast<short>(width), static_cast<short>(height) };
+    RETURN_IF_FAILED(_renderEngine->SetWindowSize(windowSize));
+
+    // Invalidate everything
+    _renderer->TriggerRedrawAll();
+
+    // Convert our new dimensions to characters
+    const auto viewInPixels = Viewport::FromDimensions({ 0, 0 },
+                                                       { static_cast<short>(windowSize.cx), static_cast<short>(windowSize.cy) });
+    const auto vp = _renderEngine->GetViewportInCharacters(viewInPixels);
+
+    // If this function succeeds with S_FALSE, then the terminal didn't
+    //      actually change size. No need to notify the connection of this
+    //      no-op.
+    // TODO: MSFT:20642295 Resizing the buffer will corrupt it
+    // I believe we'll need support for CSI 2J, and additionally I think
+    //      we're resetting the viewport to the top
+    RETURN_IF_FAILED(_terminal->UserResize({ vp.Width(), vp.Height() }));

```

Comment:
More suspected copy paste. I don't like this because it's clear that the code it's coming from is interim and slated to be updated and this copy will be forgotten/lost.